NuRadioReco/modules/channelCWNotchFilter.py

```diff
@@ -0,0 +1,227 @@
+    """
+    Function fo find the frequency peaks in the real fourier transform of the input trace.
+
+    Parameters
+    ----------
+    trace : np.ndarray
+        Waveform 
+    fs : float
+        Sampling frequency, (input should be taking from the channel object)
+    threshold : float, default = 4
+        Threshold for peak definition. A peak is defined as a point in the frequency spectrum
+        that exceeds threshold * rms(real fourier transform)
+    
+    Returns
+    -------
+    freq_peaks : np.ndarray
+        Frequencies at which a peak was found
+    """
+    freq = np.fft.rfftfreq(len(trace), d=1/fs)
+    ft = fft.time2freq(trace, fs)
+    
+    freq_peaks = find_frequency_peaks(freq, ft, fs=fs, threshold=threshold)
+
+    return freq_peaks
+
+def find_frequency_peaks(freq: np.ndarray, spectrum : np.ndarray, threshold : float = 4):
+    """
+    Function fo find the frequency peaks in the real fourier transform of the input trace.
+
+    Parameters
+    ----------
+    freq : np.ndarray
+        Frequencies of a NuRadio time trace
+    spectrum : np.ndarray
+        Spectrum of a NuRadio time trace
+    threshold : float, default = 4
+        Threshold for peak definition. A peak is defined as a point in the frequency spectrum
+        that exceeds threshold * rms(real fourier transform)
+    
+    Returns
+    -------
+    freq : np.ndarray
+        Frequencies at which a peak was found
+    """
+    
+    rms = np.sqrt(np.mean(np.abs(spectrum)**2))
+    peak_idxs = np.where(np.abs(spectrum) > threshold * rms)[0]
+
+    return freq[peak_idxs]
+
+
+def filter_cws(trace : np.ndarray, freq : np.ndarray, spectrum : np.ndarray, fs=3.2e9 * units.Hz, quality_factor=1e3, threshold=4):
+    """
+    Function that applies a notch filter at the frequency peaks of a given time trace
+    using the scipy library
+
+    Parameters
+    ----------
+    trace : np.ndarray
+        waveform (shape: [2048])
+    freq : np.ndarray
+        Frequency of the trace's real fourier transform
+    spectrum:
+        the trace's real fourier transform
+    fs : float, default = 3.2e9 Hz
+        sampling frequency of the RNO-G DAQ
+    quality_factor : int, default = 1000
+        quality factor of the notch filter, defined as the ratio f0/bw, where f0 is the centre frequency
+        and bw the bandwidth of the filter at (f0,-3 dB)
+    threshold : int, default = 4
+        threshold for peak definition. A peak is defined as a point in the frequency spectrum
+        that exceeds threshold * rms(real fourier transform)
+
+    """
+    freqs = find_frequency_peaks(freq, spectrum, threshold=threshold)
+
+    if len(freqs):
+        notch_filters = [signal.iirnotch(freq, quality_factor, fs = fs) for freq in freqs]
+        trace_notched = signal.filtfilt(notch_filters[0][0], notch_filters[0][1], trace)
+        for notch in notch_filters[1:]:
+            trace_notched = signal.filtfilt(notch[0], notch[1], trace_notched)
```

Comment:
I don't know how filtfilt is implemented, but depending on implementation it could make sense to cascade the filters first and only apply once? Though that could be numerically unstable, depending on implementation (I think, what you'd want to do is convert all of the filters into second-order sections and then cascade? I know in MATLAB filtfilt supports filters in second-order section form as input but I don't know how it works in scipy at all, unfortunately. There probably should be away to use second-order sections ('sos') otherwise you run into numerical stability problems at high orders anyway?)

Reply:
Right, a 0-phase linear filter (other than the trivial one) won't be causal (for example, a brick wall/sinc filter has 0 phase, and has a really acausal response, but in general this should follow from the Kramers-Kronig relations). So this means it could look ugly when plotted and affect time-domain metrics like impulsivity, but on the other hand it has good properties for cross-correlation (it basically just acts as a frequency weighting for cross-correlation, so it mostly just changes the normalization for cross-correlation). Crucially, this won't bias the inter-channel lag measurements (very important if you're applying different filters to different channels), so I guess a 0-phase filter has to be the right choice here! 

I wonder if it makes sense to optionally return the applied filter as well so that it also could be applied to e.g. a template used for cross-correlation, or as some sort of correction to an impulsivity measurement? This could be implemented by having an optional kwarg that is either None (default) or a list-like that the filter coefficients could be appended to.